torch/csrc/jit/passes/mkldnn_rewrite.cpp

```diff
@@ -6,13 +6,50 @@
 #include <torch/csrc/jit/passes/dead_code_elimination.h>
 #include <torch/csrc/jit/passes/graph_rewrite_helper.h>
 #include <torch/csrc/jit/passes/mkldnn_rewrite.h>
+#include <torch/csrc/jit/passes/mkldnn_rewrite_helper.h>
 #include <torch/csrc/jit/tensorexpr/kernel.h>
 
 namespace torch {
 namespace jit {
 
 #if AT_MKLDNN_ENABLED()
 
+namespace mkldnn {
+
+const std::vector<std::string> zero_scalar_operand =
+    std::vector<std::string>({});
+const std::string algorithm_indicator = std::string("%algorithm");
+
+std::string construct_operand_list(
+    std::vector<std::string> scalar_input,
+    std::string algorithm_indicator) {
+  std::string constructed_operand_list = "";
+
+  std::string joined_scalar_operands = c10::Join(", ", scalar_input);
+  std::string scalar_operands = "%scalars : Scalar?[] = prim::ListConstruct(" +
+      joined_scalar_operands + ")\n";
+
+  constructed_operand_list += scalar_operands;
+
+  if (algorithm_indicator.empty()) {
+    std::string algorithm_placeholder = "%algorithm : str? = prim::Constant()";
```

Comment:
I suspect there is a "\n" missing somewhere here. Do we have tests exercising this code?

Reply:
It is tested by `test_conv_eltwise` in `test/test_mkldnn_fusion.py`:
https://github.com/pytorch/pytorch/blob/7df0878b9936961cc1bde9d20c834ac4331d140a/test/test_mkldnn_fusion.py#L87
`\n` is not necessarily needed. The returned value of this function will be used as the `construct_operand_list` in the below CodeTemplate [here](https://github.com/pytorch/pytorch/blob/gh/XiaobingSuper/10/head/torch/csrc/jit/passes/mkldnn_rewrite.cpp#L196-L204
), which has already implicitly specified the `\n` after `${construct_operand_list}`:

```cpp
graph(${graph_input}
      %input_size:int[], %attr_placeholder:str, %scalars_placeholder: Scalar?[], %algorithm_placeholder: str?${op_input_str}):
    %attr: str = prim::Constant[value="${op_attr}"]()
    ${construct_operand_list}
    %packed_weight_bias : __torch__.torch.classes.${op_context_name} =  ${prepack_op_name}(
        ${prepack_input}
        %input_size, %attr, %scalars, %algorithm)
    %res = ${run_op_name}(%input, %packed_weight_bias)
    return (%res))");
```


The IR with the current code is:
```bash
%16 : str? = prim::Constant()
%17 : __torch__.torch.classes.mkldnn.ConvOpContext = mkldnn_prepacked::conv2d_prepack(%self.conv.weight, %self.conv.bias, %3, %2, %3, %1, %8, %14, %15, %16)
```

Reply:
It still is desirable for code readability to stay consistent in whether we end the string with the new line or not.

Reply:
Also, I don't see how that test exercises different algorithms. Could you please point me to that place?

Reply:
@ZolotukhinM . I clean the code to avoid the confusion.